File: top.js
const { log } = require("./lib/logging");
const { delay } = require("./lib/util");
const { loadCookies, saveCookies, clearCookies } = require("./lib/cookies");
const { connectBrowser } = require("./lib/browser");
const { isLoggedInToTopgg, doOAuth } = require("./lib/oauth");
const { voteForBot } = require("./lib/vote");

// Main entry point: one session per token, votes all bots in that session.
// `label` is an optional human-readable name for logging (falls back to a
// truncated token prefix if not provided).
async function voteAllBotsForToken(token, botIds, label) {
  const shortT = label || (token.slice(0, 5) + "...");

  log(`[token:${shortT}] Validating...`);
  try {
    const res = await fetch("https://discord.com/api/v10/users/@me", {
      headers: { Authorization: token },
    });
    if (!res.ok) {
      log(`[token:${shortT}] Invalid token, skipping`);
      clearCookies(token);
      const failResults = {};
      for (const botId of botIds) failResults[botId] = "failed";
      return failResults;
    }
  } catch (err) {
    log(`[token:${shortT}] Validation error: ${err.message}`);
    const failResults = {};
    for (const botId of botIds) failResults[botId] = "failed";
    return failResults;
  }
  log(`[token:${shortT}] Valid`);

  const results = {}; // botId -> "confirmed" | "already" | "turnstile" | "failed"

  let browser, context, page;
  try {
    ({ browser, context, page } = await connectBrowser());
    log(`[token:${shortT}] Browser connected`);

    const savedCookies = loadCookies(token);
    let loggedIn = false;

    if (savedCookies) {
      log(`[token:${shortT}] Loading saved cookies...`);
      await context.addCookies(savedCookies);
      loggedIn = await isLoggedInToTopgg(page);
      if (loggedIn) {
        log(`[token:${shortT}] Cookie login successful`);
      } else {
        log(`[token:${shortT}] Cookies expired, re-authing...`);
        clearCookies(token);
      }
    }

    if (!loggedIn) {
      await doOAuth(page, token, shortT);
      const cookies = await context.cookies();
      saveCookies(token, cookies);
      log(`[token:${shortT}] Cookies saved (${cookies.length} total): ${cookies.map(c => `${c.name}[domain=${c.domain},path=${c.path},secure=${c.secure},sameSite=${c.sameSite}]`).join(", ")}`);
    }

    // Sequential processing, one bot/page at a time, reusing the single
    // `page` throughout. Pre-opening a tab per bot (to overlap the ~10s ad
    // countdown / Turnstile render with the current bot's processing) was
    // tried and reverted: browserless-v2 is on a 1GB memory limit, and 3
    // concurrent tabs (shared page + 2 pre-opened) already pushed it to
    // ~721MB+ in one run — confirmed via the Railway metrics dashboard,
    // with the memory/CPU spike landing at the exact timestamp of a bot #1
    // goto() timeout and a bot #3 3-minute stall in waitForVoteStatus. That
    // headroom is too thin to risk an OOM kill for a ~10-20s/bot speedup.
    for (const botId of botIds) {
      try {
        results[botId] = await voteForBot(page, shortT, botId);
      } catch (err) {
        log(`[token:${shortT}] Error voting for bot ${botId}: ${err.message}`);
        results[botId] = "failed";
        if (/page crashed/i.test(err.message) || /target crashed/i.test(err.message)) {
          log(`[token:${shortT}] Browser/page crashed — aborting remaining bots for this session`);
          break;
        }
      }
      await delay(2000);
    }

    for (const botId of botIds) {
      if (!(botId in results)) results[botId] = "failed";
    }

    const updatedCookies = await context.cookies().catch(() => null);
    if (updatedCookies) saveCookies(token, updatedCookies);

  } catch (err) {
    log(`[token:${shortT}] Fatal error: ${err.message}`);
    for (const botId of botIds) {
      if (!(botId in results)) results[botId] = "failed";
    }
  } finally {
    if (browser) {
      try { await browser.close(); } catch (_) {}
      log(`[token:${shortT}] Browser closed`);
    }
  }

  return results;
}

module.exports = { voteAllBotsForToken };

File: top.py
"""Port of top.js — one session per token, votes all bots in that session."""
import aiohttp

from lib.logging import log, log_verbose
from lib.util import delay
from lib.cookies import load_cookies, save_cookies, clear_cookies
from lib.browser import connect_browser, prepare_page
from lib.oauth import is_logged_in_to_topgg, do_oauth
from lib.vote import vote_for_bot


async def vote_all_bots_for_token(playwright, token, bot_ids, label=None) -> dict:
    short_t = label or (token[:5] + "...")

    log(f"[token:{short_t}] Validating...")
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(
                "https://discord.com/api/v10/users/@me",
                headers={"Authorization": token},
            ) as res:
                if not res.ok:
                    log(f"[token:{short_t}] Invalid token, skipping")
                    clear_cookies(token)
                    return {bot_id: "failed" for bot_id in bot_ids}
    except Exception as err:
        log(f"[token:{short_t}] Validation error: {err}")
        return {bot_id: "failed" for bot_id in bot_ids}
    log(f"[token:{short_t}] Valid")

    results = {}  # bot_id -> "confirmed" | "already" | "turnstile" | "failed"

    browser = None
    try:
        browser, context, page = await connect_browser(playwright)
        log(f"[token:{short_t}] Browser connected")

        saved_cookies = load_cookies(token)
        logged_in = False

        if saved_cookies:
            log(f"[token:{short_t}] Loading saved cookies...")
            await context.add_cookies(saved_cookies)
            logged_in = await is_logged_in_to_topgg(page)
            if logged_in:
                log(f"[token:{short_t}] Cookie login successful")
            else:
                log(f"[token:{short_t}] Cookies expired, re-authing...")
                clear_cookies(token)

        if not logged_in:
            await do_oauth(page, token, short_t)
            cookies = await context.cookies()
            save_cookies(token, cookies)
            log(f"[token:{short_t}] Cookies saved ({len(cookies)} total)")
            log_verbose(f"[token:{short_t}] Cookie detail: " + ", ".join(
                f"{c['name']}[domain={c.get('domain')},path={c.get('path')},secure={c.get('secure')},sameSite={c.get('sameSite')}]"
                for c in cookies
            ))

        # Sequential processing, one bot/page at a time, reusing the single
        # `page` throughout — same reasoning as the original: concurrent
        # tabs previously pushed the browser process too close to its
        # memory limit for the modest speedup gained.
        for bot_id in bot_ids:
            try:
                results[bot_id] = await vote_for_bot(page, short_t, bot_id)
            except Exception as err:
                log(f"[token:{short_t}] Error voting for bot {bot_id}: {err}")
                results[bot_id] = "failed"
                msg = str(err)
                if "page crashed" in msg.lower() or "target crashed" in msg.lower():
                    log(f"[token:{short_t}] Browser/page crashed — aborting remaining bots for this session")
                    break
            await delay(2000)

        for bot_id in bot_ids:
            if bot_id not in results:
                results[bot_id] = "failed"

        try:
            updated_cookies = await context.cookies()
            save_cookies(token, updated_cookies)
        except Exception:
            pass

    except Exception as err:
        log(f"[token:{short_t}] Fatal error: {err}")
        for bot_id in bot_ids:
            if bot_id not in results:
                results[bot_id] = "failed"
    finally:
        if browser:
            try:
                await browser.close()
            except Exception:
                pass
            log(f"[token:{short_t}] Browser closed")

    return results
